clic/src/cleCloseIndexGapsInLabelMapKernel.cpp

```diff
@@ -0,0 +1,112 @@
+/*  CLIc - version 0.1 - Copyright 2020 Stéphane Rigaud, Robert Haase,
+*   Institut Pasteur Paris, Max Planck Institute for Molecular Cell Biology and Genetics Dresden
+*
+*   CLIc is part of the clEsperanto project http://clesperanto.net 
+*
+*   This file is subject to the terms and conditions defined in
+*   file 'LICENSE.txt', which is part of this source code package.
+*/
+
+
+#include "cleCloseIndexGapsInLabelMapKernel.h"
+
+#include "cleMaximumOfAllPixelsKernel.h"
+#include "cleFlagExistingLabelsKernel.h"
+#include "cleSetKernel.h"
+#include "cleSumReductionKernel.h"
+#include "cleSetColumnKernel.h"
+#include "cleBlockEnumerateKernel.h"
+#include "cleReplaceIntensitiesKernel.h"
+
+
+namespace cle
+{
+
+void CloseIndexGapsInLabelMapKernel::SetInput(Object& x)
+{
+    this->AddObject(&x, "src");
+}
+
+void CloseIndexGapsInLabelMapKernel::SetOutput(Object& x)
+{
+    this->AddObject(&x, "dst");
+}
+
+void CloseIndexGapsInLabelMapKernel::SetBlockSize(int x)
+{
+    this->blocksize = x;
+}
+
+void CloseIndexGapsInLabelMapKernel::Execute()
+{
+    Buffer* src = dynamic_cast<Buffer*>(parameterList.at("src"));
+    Buffer* dst = dynamic_cast<Buffer*>(parameterList.at("dst"));
+
+    size_t arr_size = src->GetDimensions()[0] * src->GetDimensions()[1] * src->GetDimensions()[2];
+    unsigned int one_dim[3] = {1,1,1};
+
+    cl_mem max_mem = CreateBuffer<float>(sizeof(float), this->gpu);
+    Buffer* max  = new Buffer(max_mem, one_dim, src->GetDataType());
+
+    MaximumOfAllPixelsKernel maximumOfPixels(this->gpu);
+    maximumOfPixels.SetInput(*src);
+    maximumOfPixels.SetOutput(*max);
+    maximumOfPixels.Execute();
+
+    float max_value = ReadBuffer<float>(max->GetData(), max->GetBitSize() * 1, this->gpu)[0];
+    unsigned int nb_indices =  int(max_value) + 1;
+
+    unsigned int indices_dim[3] = {nb_indices, 1, 1};
+    cl_mem flaggedIndices_mem = CreateBuffer<float>(src->GetBitSize() * indices_dim[0], this->gpu);
+    Buffer* flaggedIndices  = new Buffer(flaggedIndices_mem, indices_dim, src->GetDataType());
+
+    SetKernel set(this->gpu);
+    set.SetInput(*flaggedIndices);
+    set.SetValue(0);
+    set.Execute();
```

Comment:
Setting this image to 0 should be done inside the FlagExistingLabelsKernel. We can then spare this step here (se commend further down)

Reply:
That's btw. also "wrong" in pyclesperanto. Here we see how code quality benefits from rewriting things and reviewing thm :-) I will fix this on pyclesperanto side.

Reply:
`SetKernel` is removed.